File: backend/internal/model/owner.go
package model

import (
	"encoding/json"
	"time"
)

type OwnerRole struct {
	ID                  int        `json:"id"`
	TelegramUserID      int64      `json:"telegram_user_id"`
	Role                string     `json:"role"` // 'super_admin', 'admin', 'moderator', 'support'
	TotpSecret          string     `json:"-"`
	TotpEnabled         bool       `json:"totp_enabled"`
	TotpEnabledAt       *time.Time `json:"totp_enabled_at,omitempty"`
	RecoveryCodesHashes []string   `json:"-"`
	PasswordHash        string     `json:"-"`
	IPWhitelist         []string   `json:"ip_whitelist"`
	CreatedAt           time.Time  `json:"created_at"`
	LastLoginAt         *time.Time `json:"last_login_at"`
}

type OwnerAuditLog struct {
	ID           int64           `json:"id"`
	OwnerID      int64           `json:"owner_id"`
	Action       string          `json:"action"`
	TargetUserID *int64          `json:"target_user_id,omitempty"`
	TargetID     string          `json:"target_id,omitempty"`
	Payload      json.RawMessage `json:"payload,omitempty"`
	IPAddress    string          `json:"ip_address,omitempty"`
	UserAgent    string          `json:"user_agent,omitempty"`
	CreatedAt    time.Time       `json:"created_at"`
}

type UserBan struct {
	UserID    int64      `json:"user_id"`
	BanType   string     `json:"ban_type"` // 'full', 'shadow', 'wallet_freeze'
	Reason    string     `json:"reason,omitempty"`
	BannedBy  int64      `json:"banned_by"`
	BannedAt  time.Time  `json:"banned_at"`
	ExpiresAt *time.Time `json:"expires_at,omitempty"`
}

type ImpersonationSession struct {
	ID              string          `json:"id"`
	OwnerID         int64           `json:"owner_id"`
	TargetUserID    int64           `json:"target_user_id"`
	StartedAt       time.Time       `json:"started_at"`
	EndedAt         *time.Time      `json:"ended_at,omitempty"`
	DurationSeconds int             `json:"duration_seconds,omitempty"`
	ActionsTaken    json.RawMessage `json:"actions_taken,omitempty"`
}

type ChartPoint struct {
	Date  string  `json:"date"`
	Value float64 `json:"value"`
}

type TodayEconomy struct {
	MintedToday       float64 `json:"minted_today"`
	BurnedToday       float64 `json:"burned_today"`
	DecayedToday      float64 `json:"decayed_today"`
	RevSharePaidToday float64 `json:"rev_share_paid_today"`
}

type OwnerDashboardStats struct {
	DAU              int             `json:"dau"`
	MAU              int             `json:"mau"`
	TotalUsers       int             `json:"total_users"`
	FrgCirculation   float64         `json:"frg_circulation"`
	CoinsCirculation float64         `json:"coins_circulation"`
	StarsVolume      float64         `json:"stars_volume"`
	DauTrend         float64         `json:"dau_trend"`
	MauTrend         float64         `json:"mau_trend"`
	RevenueTrend     float64         `json:"revenue_trend"`
	CirculationTrend float64         `json:"circulation_trend"`
	TodayEconomy     TodayEconomy    `json:"today_economy"`
	TotpEnabled      bool            `json:"totp_enabled"`
	TotpGraceDays    int             `json:"totp_grace_days_left"`
	RecentActivity   []OwnerAuditLog `json:"recent_activity"`
	DauChart         []ChartPoint    `json:"dau_chart"`
	CoinFlowChart    []ChartPoint    `json:"coin_flow_chart"`
	RecentSignups    []SearchedUser  `json:"recent_signups,omitempty"`
}

type SearchedUser struct {
	TelegramID   int64     `json:"telegram_id"`
	Username     string    `json:"username"`
	FirstName    string    `json:"first_name"`
	LastName     string    `json:"last_name"`
	LanguageCode string    `json:"language_code"`
	Balance      float64   `json:"balance"`
	IsPremium    bool      `json:"is_premium"`
	IsFlagged    bool      `json:"is_flagged"`
	FraudReason  string    `json:"fraud_reason,omitempty"`
	IsBanned     bool      `json:"is_banned"`
	BanType      string    `json:"ban_type,omitempty"`
	BanReason    string    `json:"ban_reason,omitempty"`
	BanExpiresAt *time.Time `json:"ban_expires_at,omitempty"`
	CreatedAt    time.Time `json:"created_at"`
}

type PromoCode struct {
	Code         string     `json:"code"`
	RewardAmount float64    `json:"reward_amount"`
	MaxUses      int        `json:"max_uses"`
	UsesCount    int        `json:"uses_count"`
	ExpiresAt    *time.Time `json:"expires_at,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
}

type PromoRedemption struct {
	ID         int       `json:"id"`
	Code       string    `json:"code"`
	UserID     int64     `json:"user_id"`
	RedeemedAt time.Time `json:"redeemed_at"`
}

type ManagedUserbot struct {
	ID            string    `json:"id"`
	PhoneNumber   string    `json:"phone_number"`
	Status        string    `json:"status"`
	ChannelsCount int       `json:"channels_count"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

type DashboardAd struct {
	ID        string     `json:"id"`
	Slot      string     `json:"slot,omitempty"`
	Title     string     `json:"title"`
	AltText   string     `json:"alt_text,omitempty"`
	ImageURL  string     `json:"image_url"`
	TargetURL string     `json:"target_url,omitempty"`
	Target    string     `json:"target,omitempty"` // legacy alias
	IsActive  bool       `json:"is_active"`
	Priority  int        `json:"priority,omitempty"`
	StartDate *time.Time `json:"start_date,omitempty"`
	EndDate   *time.Time `json:"end_date,omitempty"`
}

type AdCampaign struct {
	ID               string     `json:"id"`
	Slot             string     `json:"slot"`
	Title            string     `json:"title"`
	AltText          string     `json:"alt_text"`
	ImageURL         string     `json:"image_url"`
	TargetURL        string     `json:"target_url"`
	IsActive         bool       `json:"is_active"`
	Priority         int        `json:"priority"`
	StartDate        *time.Time `json:"start_date,omitempty"`
	EndDate          *time.Time `json:"end_date,omitempty"`
	ImpressionsCount int64      `json:"impressions_count"`
	ClicksCount      int64      `json:"clicks_count"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
}

type SystemSettings struct {
	MaintenanceMode      bool          `json:"maintenance_mode"`
	TapMultiplier        float64       `json:"tap_multiplier"`
	ReferralBonus        int           `json:"referral_bonus"`
	DailyRewardBase      int           `json:"daily_reward_base"`
	FatigueThreshold1    int           `json:"fatigue_threshold_1,omitempty"`
	FatigueThreshold2    int           `json:"fatigue_threshold_2,omitempty"`
	FatigueThreshold3    int           `json:"fatigue_threshold_3,omitempty"`
	TapBotCapSeconds     int           `json:"tap_bot_cap_seconds,omitempty"`
	ReferralRevSharePct  float64       `json:"referral_rev_share_pct,omitempty"`
	CoinDecayPct         float64       `json:"coin_decay_pct,omitempty"`
	CoinExpiryDays       int           `json:"coin_expiry_days,omitempty"`
	TurboDurationSeconds int           `json:"turbo_duration_seconds,omitempty"`
	InflationCap         float64       `json:"inflation_cap,omitempty"`
	DashboardAds         []DashboardAd `json:"dashboard_ads,omitempty"`
	Version              int           `json:"version"`
}

type Broadcast struct {
	ID             string     `json:"id"`
	OwnerID        int64      `json:"owner_id"`
	TargetAudience string     `json:"target_audience"` // 'all', 'premium', 'active_7d', 'inactive'
	Message        string     `json:"message"`
	Status         string     `json:"status"` // 'draft', 'scheduled', 'sending', 'completed', 'failed', 'paused'
	ScheduledAt    *time.Time `json:"scheduled_at,omitempty"`
	SentCount      int        `json:"sent_count"`
	TotalCount     int        `json:"total_count"`
	FailedCount    int        `json:"failed_count"`
	StartedAt      *time.Time `json:"started_at,omitempty"`
	CompletedAt    *time.Time `json:"completed_at,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
}

type OrderRecord struct {
	ID        string    `json:"id"`
	UserID    int64     `json:"user_id"`
	Amount    int       `json:"amount"`
	Status    string    `json:"status"`
	Payload   string    `json:"payload"`
	CreatedAt time.Time `json:"created_at"`
}

type FinanceSummary struct {
	TotalRevenueStars   int64   `json:"total_revenue_stars"`
	Revenue7d           int64   `json:"revenue_7d"`
	Revenue30d          int64   `json:"revenue_30d"`
	TotalOrders         int64   `json:"total_orders"`
	ActiveSubscriptions int64   `json:"active_subscriptions"`
	ChurnRate           float64 `json:"churn_rate"`
}

type PremiumEntity struct {
	EntityType    string     `json:"entity_type"` // "channel" or "group"
	EntityID      string     `json:"entity_id"`
	Title         string     `json:"title"`
	OwnerID       int64      `json:"owner_id"`
	CreditBalance float64    `json:"credit_balance"`
	PremiumUntil  *time.Time `json:"premium_until"`
}

type SystemErrorLog struct {
	ID           string    `json:"id"`
	Source       string    `json:"source"`
	ErrorMessage string    `json:"error_message"`
	Level        string    `json:"level,omitempty"` // "error", "warn", "info"
	CreatedAt    time.Time `json:"created_at"`
}

type SystemHealthMetrics struct {
	DBStatus          string `json:"db_status"`
	DBLatencyMS       int64  `json:"db_latency_ms"`
	RedisStatus       string `json:"redis_status"`
	ActiveGoroutines  int    `json:"active_goroutines"`
	MemoryUsedMB      uint64 `json:"memory_used_mb"`
	AllocatedMB       uint64 `json:"allocated_mb"`
	TotalSysMB        uint64 `json:"total_sys_mb"`
	CPUUsagePercent   float64 `json:"cpu_usage_percent"`
	UptimeSeconds     int64  `json:"uptime_seconds"`
	RecentErrorsCount int    `json:"recent_errors_count"`
	Goroutines        int    `json:"goroutines,omitempty"`
}

type EntityRecord struct {
	ID            string     `json:"id"`
	EntityType    string     `json:"entity_type"`
	EntityID      string     `json:"entity_id"`
	Title         string     `json:"title"`
	Status        string     `json:"status"`
	OwnerID       int64      `json:"owner_id"`
	OwnerUsername string     `json:"owner_username,omitempty"`
	CreditBalance float64    `json:"credit_balance"`
	PaidUntil     *time.Time `json:"paid_until,omitempty"`
}

File: frontend/src/shared/api/owner.ts
import { apiClient } from './axios.js';

/* ==========================================================================
   Type Definitions for Owner Domain Entities
   ========================================================================== */

export interface SearchedUser {
	telegram_id: number;
	username: string;
	first_name: string;
	last_name: string;
	language_code: string;
	created_at: string;
	balance: number;
	is_premium: boolean;
	is_flagged: boolean;
	fraud_reason: string;
	is_banned?: boolean;
	ban_type?: string;
	ban_reason?: string;
	ban_expires_at?: string;
}

export interface AuditLogEntry {
	id: string | number;
	owner_id: number;
	target_type?: string;
	target_id?: string | number;
	action: string;
	ip_address?: string;
	user_agent?: string;
	payload?: Record<string, any>;
	created_at: string;
}

export interface QuestItem {
	id: string | number;
	title: string;
	description: string;
	reward_frg: number;
	reward_xp: number;
	type: 'telegram_channel' | 'telegram_group' | 'daily_checkin' | 'invite' | 'external_link' | 'partner';
	is_active: boolean;
	config?: Record<string, any>;
	expires_at?: string;
	parent_id?: string | number;
	created_at?: string;
}

export interface AdminDailyCombo {
	id: number;
	active_date: string;
	secret_word: string;
	reward_amount: number;
	created_by?: string;
}

export interface ManagedUserbot {
	id: string;
	phone_number: string;
	status: 'connected' | 'connecting' | 'expired' | 'error';
	channels_count: number;
	created_at: string;
	updated_at: string;
}

export interface SystemSettings {
	maintenance_mode: boolean;
	tap_multiplier: number;
	referral_bonus: number;
	daily_reward_base: number;
	dashboard_ads?: DashboardAd[];
}

export interface DashboardAd {
	id: string;
	title: string;
	image_url: string;
	target: string;
	is_active: boolean;
	start_date?: string;
	end_date?: string;
}

export interface PromoCode {
	id: string | number;
	code: string;
	reward_frg: number;
	max_uses: number;
	current_uses: number;
	expires_at: string;
	is_active: boolean;
	created_at: string;
}

export interface BroadcastMessage {
	id: string | number;
	target_audience: 'all' | 'premium' | 'active_7d' | 'inactive';
	message_text: string;
	status: 'draft' | 'scheduled' | 'sending' | 'completed' | 'failed';
	scheduled_at?: string;
	sent_count: number;
	total_count: number;
	created_at: string;
}

export interface FinanceOrder {
	id: string;
	user_id: number;
	username?: string;
	amount_stars: number;
	status: 'paid' | 'pending' | 'failed' | 'refunded';
	item_type: string;
	created_at: string;
}

export interface SystemHealthMetrics {
	db_status: 'ok' | 'degraded' | 'down';
	db_latency_ms: number;
	redis_status: 'ok' | 'down';
	active_goroutines: number;
	memory_used_mb: number;
	uptime_seconds: number;
	recent_errors_count: number;
}

export interface OwnerEntityItem {
	id: string | number;
	type: 'channel' | 'group';
	title: string;
	username?: string;
	telegram_id: number;
	owner_id: number;
	owner_username?: string;
	credit_balance: number;
	status: 'active' | 'suspended' | 'expired';
	created_at: string;
}

/* ==========================================================================
   Unified Typed Owner API Client
   ========================================================================== */

export const ownerApi = {
	// --- Users ---
	searchUsers: (query: string) =>
		apiClient.get<SearchedUser[]>(`/owner/users/search?q=${encodeURIComponent(query)}`).then((r) => r.data),

	adjustFrg: (userId: number, amount: number, reason: string) =>
		apiClient
			.post<{ success: boolean; new_balance: number }>('/owner/users/adjust-frg', {
				user_id: userId,
				amount,
				reason,
			})
			.then((r) => r.data),

	banUser: (userId: number, banType: string, reason: string, durationSeconds: number) =>
		apiClient
			.post<{ success: boolean }>('/owner/users/ban', {
				user_id: userId,
				ban_type: banType,
				reason,
				duration_seconds: durationSeconds,
			})
			.then((r) => r.data),

	unbanUser: (userId: number) =>
		apiClient.post<{ success: boolean }>('/owner/users/unban', { user_id: userId }).then((r) => r.data),

	flagUser: (userId: number, isFlagged: boolean, fraudReason: string) =>
		apiClient
			.post<{ success: boolean }>('/owner/users/flag', {
				user_id: userId,
				is_flagged: isFlagged,
				fraud_reason: fraudReason,
			})
			.then((r) => r.data),

	impersonateUser: (userId: number) =>
		apiClient.post<{ token: string }>('/owner/users/impersonate', { user_id: userId }).then((r) => r.data),

	// --- Audit Logs ---
	getAuditLogs: (limit = 20, offset = 0) =>
		apiClient
			.get<{ logs: AuditLogEntry[]; has_more: boolean }>(`/owner/audit-logs?limit=${limit}&offset=${offset}`)
			.then((r) => r.data),

	// --- Quests ---
	listQuests: () => apiClient.get<QuestItem[]>('/owner/quests').then((r) => r.data),
	createQuest: (quest: Partial<QuestItem>) => apiClient.post<QuestItem>('/owner/quests', quest).then((r) => r.data),
	updateQuest: (id: string | number, quest: Partial<QuestItem>) =>
		apiClient.put<QuestItem>(`/owner/quests/${id}`, quest).then((r) => r.data),
	deleteQuest: (id: string | number) => apiClient.delete(`/owner/quests/${id}`).then((r) => r.data),

	// --- Daily Combos ---
	listCombos: () => apiClient.get<AdminDailyCombo[]>('/owner/combos').then((r) => r.data),
	createCombo: (date: string, word: string, reward: number) =>
		apiClient
			.post<AdminDailyCombo>('/owner/combos', { date, secret_word: word, reward_amount: reward })
			.then((r) => r.data),
	deleteCombo: (id: number) => apiClient.delete(`/owner/combos/${id}`).then((r) => r.data),

	// --- Userbots ---
	listUserbots: () => apiClient.get<ManagedUserbot[]>('/owner/userbots').then((r) => r.data),
	sendUserbotCode: (phone: string) =>
		apiClient.post<{ phone_code_hash: string }>('/owner/userbot/send-code', { phone }).then((r) => r.data),
	verifyUserbotCode: (phone: string, code: string, phoneCodeHash: string, password_2fa?: string) =>
		apiClient
			.post('/owner/userbot/verify-code', { phone, code, phone_code_hash: phoneCodeHash, password_2fa })
			.then((r) => r.data),
	deleteUserbot: (id: string) => apiClient.delete(`/owner/userbots/${id}`).then((r) => r.data),

	// --- System Settings & Ads (Decoupled Payload Safety) ---
	getSettings: () => apiClient.get<SystemSettings>('/owner/settings').then((r) => r.data),
	updateSettings: (settingsPayload: SystemSettings) =>
		apiClient.put<SystemSettings>('/owner/settings', settingsPayload).then((r) => r.data),
	getAds: async (): Promise<DashboardAd[]> => {
		const res = await apiClient.get<SystemSettings>('/owner/settings');
		return res.data?.dashboard_ads || [];
	},
	updateAds: async (ads: DashboardAd[]) => {
		const current = await apiClient.get<SystemSettings>('/owner/settings');
		return (
			await apiClient.put<SystemSettings>('/owner/settings', {
				...current.data,
				dashboard_ads: ads,
			})
		).data;
	},

	// --- Promos ---
	listPromos: () => apiClient.get<PromoCode[]>('/owner/promos').then((r) => r.data),
	createPromo: (code: string, reward: number, maxUses: number, expiryDate: string) =>
		apiClient
			.post<PromoCode>('/owner/promos', {
				code,
				reward_frg: reward,
				max_uses: maxUses,
				expires_at: expiryDate,
			})
			.then((r) => r.data),
	deletePromo: (id: string | number) => apiClient.delete(`/owner/promos/${id}`).then((r) => r.data),

	// --- Broadcast ---
	listBroadcasts: () =>
		apiClient
			.get<BroadcastMessage[]>('/owner/broadcast')
			.then((r) => r.data)
			.catch(() => []),
	sendBroadcast: (audience: string, text: string, scheduledAt?: string) =>
		apiClient
			.post<{ success: boolean; id: string | number }>('/owner/broadcast', {
				target_audience: audience,
				message_text: text,
				scheduled_at: scheduledAt,
			})
			.then((r) => r.data),

	// --- Finance ---
	getFinanceOrders: () => apiClient.get<FinanceOrder[]>('/owner/finance/orders').then((r) => r.data),

	// --- Health ---
	getHealthMetrics: () => apiClient.get<SystemHealthMetrics>('/owner/health/metrics').then((r) => r.data),
	getHealthLogs: () =>
		apiClient
			.get<any[]>('/owner/health/errors')
			.then((r) => (Array.isArray(r.data) ? r.data.map((e) => `[${e.source || 'SYS'}] ${e.error_message || ''}`) : []))
			.catch(() => []),

	// --- Entities ---
	listEntities: async (): Promise<OwnerEntityItem[]> => {
		const [channelsRes, groupsRes] = await Promise.all([
			apiClient.get<any[]>('/owner/entities/channels').catch(() => ({ data: [] })),
			apiClient.get<any[]>('/owner/entities/groups').catch(() => ({ data: [] })),
		]);
		const channels = (channelsRes.data || []).map((c: any) => ({ ...c, type: 'channel' as const }));
		const groups = (groupsRes.data || []).map((g: any) => ({ ...g, type: 'group' as const }));
		return [...channels, ...groups];
	},
	addEntityCredit: (id: string | number, amount: number, _reason: string) =>
		apiClient
			.post<{ success: boolean }>('/owner/entities/add-credit', {
				entity_id: String(id),
				days: amount,
			})
			.then((r) => r.data),
};
